d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -0,0 +1,208 @@
+  // Limit error report logging to every 10 minutes
+  private final RateLimitedLogger RATE_LIMITED_LOGGER =
+      new RateLimitedLogger(LOG, TimeUnit.MINUTES.toMillis(10), SystemClock.instance());
+  private int _totalUris = 0;
+  private int _matchedUris = 0;
+  private final DualReadLoadBalancerJmx _dualReadLoadBalancerJmx;
+
+  public UriPropertiesDualReadMonitor(DualReadLoadBalancerJmx dualReadLoadBalancerJmx) {
+    _dualReadLoadBalancerJmx = dualReadLoadBalancerJmx;
+  }
+
+  public void reportData(String clusterName, UriProperties property, boolean fromNewLb) {
+    ClusterMatchRecord cluster = _clusters.computeIfAbsent(clusterName, k -> new ClusterMatchRecord());
+
+    // Both zk and indis threads can race to report their data to the same cluster and perform operations on the total
+    // uris and matched uris counts at the same time, which can lead to incorrect results and unexpected behaviors.
+    // We need to lock per cluster when editing its ClusterMatchRecord, and release the lock after all operations done.
+    synchronized (cluster) {
+      if (fromNewLb) {
+        cluster._newLb = property;
+      } else {
+        cluster._oldLb = property;
+      }
+
+      _totalUris -= cluster._uris;
+      _matchedUris -= cluster._matched;
+
+      LOG.debug("Updated URI properties for cluster {}:\nOld LB: {}\nNew LB: {}",
+          clusterName, cluster._oldLb, cluster._newLb);
+
+      if (cluster._oldLb == null && cluster._newLb == null) {
+        _clusters.remove(clusterName);
+        updateJmxMetrics(clusterName, null);
+        return;
+      }
+
+      cluster._matched = 0;
+
+      if (cluster._oldLb == null || cluster._newLb == null) {
+        LOG.debug("Added new URI properties for {} for {} LB.", clusterName, fromNewLb ? "New" : "Old");
+
+        cluster._uris = (cluster._oldLb == null) ? cluster._newLb.Uris().size() : cluster._oldLb.Uris().size();
+        _totalUris += cluster._uris;
+
+        updateJmxMetrics(clusterName, cluster);
+        return;
+      }
+
+      cluster._uris = cluster._oldLb.Uris().size();
+      Set<URI> newLbUris = new HashSet<>(cluster._newLb.Uris());
+
+      for (URI uri : cluster._oldLb.Uris()) {
+        if (!newLbUris.remove(uri)) {
+          continue;
+        }
+
+        if (compareURI(uri, cluster._oldLb, cluster._newLb)) {
+          cluster._matched++;
+        }
+      }
+      // add the remaining unmatched URIs in newLbUris to the uri count
+      cluster._uris += newLbUris.size();
+
+      if (cluster._matched != cluster._uris) {
+        String msg = String.format("Mismatched cluster properties for %s (match score: %f, total uris: %d):"
+                + "\nOld LB: %s\nNew LB: %s",
+            clusterName, (double) cluster._matched / (double) cluster._uris, cluster._uris, cluster._oldLb,
+            cluster._newLb);
+        RATE_LIMITED_LOGGER.info(msg);
+        LOG.debug(msg);
```

Comment:
This is the kind of stuff that really should be avoided... It _always_ creates the string, even if it gets sent to DEBUG. This can create a lot of garbage. That's why the loggers take var args and lazily create the string only if logging at that level is enabled. Maybe you can make `RateLimitedLogger#logAllowed` public and do something like:
```
private static void infoOrDebugIfLimited(String msg, Object... args) {
  if (RATE_LIMITED_LOGGER.logAllowed()) {
    LOG.info(msg, args);
  } else {
    LOG.debug(msg, args);
  }
}
```
And that block becomes:
```suggestion
        infoOrDebugIfLimited(
            "Mismatched cluster properties for {} (match score: {}, total uris: {}):\nOld LB: {}\nNew LB: {}"
            clusterName, (double) cluster._matched / (double) cluster._uris, cluster._uris, cluster._oldLb,
            cluster._newLb);
```

Reply:
makes a lot sense. Updated

Reply:
We use the same trick now in the observer as well, and I really like it. Instead of losing the log statement altogether, it just gets hidden. If you turn on debug logging, you will see everything, but without duplicates.

Reply:
yeah, it's nice.